File: breezy/bzr/hashcache.py
# Copyright (C) 2005-2010 Canonical Ltd
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

# TODO: Up-front, stat all files in order and remove those which are deleted or
# out-of-date.  Don't actually re-read them until they're needed.  That ought
# to bring all the inodes into core so that future stats to them are fast, and
# it preserves the nice property that any caller will always get up-to-date
# data except in unavoidable cases.

# TODO: Perhaps return more details on the file to avoid statting it
# again: nonexistent, file type, size, etc

# TODO: Perhaps use a Python pickle instead of a text file; might be faster.


CACHE_HEADER = b"### bzr hashcache v5\n"

import os
import stat
import time

from .. import (
    atomicfile,
    errors,
    filters as _mod_filters,
    osutils,
    trace,
    )


FP_MTIME_COLUMN = 1
FP_CTIME_COLUMN = 2
FP_MODE_COLUMN = 5


class HashCache(object):
    """Cache for looking up file SHA-1.

    Files are considered to match the cached value if the fingerprint
    of the file has not changed.  This includes its mtime, ctime,
    device number, inode number, and size.  This should catch
    modifications or replacement of the file by a new one.

    This may not catch modifications that do not change the file's
    size and that occur within the resolution window of the
    timestamps.  To handle this we specifically do not cache files
    which have changed since the start of the present second, since
    they could undetectably change again.

    This scheme may fail if the machine's clock steps backwards.
    Don't do that.

    This does not canonicalize the paths passed in; that should be
    done by the caller.

    _cache
        Indexed by path, points to a two-tuple of the SHA-1 of the file.
        and its fingerprint.

    stat_count
        number of times files have been statted

    hit_count
        number of times files have been retrieved from the cache, avoiding a
        re-read

    miss_count
        number of misses (times files have been completely re-read)
    """
    needs_write = False

    def __init__(self, root, cache_file_name, mode=None,
                 content_filter_stack_provider=None):
        """Create a hash cache in base dir, and set the file mode to mode.

        :param content_filter_stack_provider: a function that takes a
            path (relative to the top of the tree) and a file-id as
            parameters and returns a stack of ContentFilters.
            If None, no content filtering is performed.
        """
        if not isinstance(root, str):
            raise ValueError("Base dir for hashcache must be text")
        self.root = root
        self.hit_count = 0
        self.miss_count = 0
        self.stat_count = 0
        self.danger_count = 0
        self.removed_count = 0
        self.update_count = 0
        self._cache = {}
        self._mode = mode
        self._cache_file_name = cache_file_name
        self._filter_provider = content_filter_stack_provider

    def cache_file_name(self):
        return self._cache_file_name

    def clear(self):
        """Discard all cached information.

        This does not reset the counters."""
        if self._cache:
            self.needs_write = True
            self._cache = {}

    def scan(self):
        """Scan all files and remove entries where the cache entry is obsolete.

        Obsolete entries are those where the file has been modified or deleted
        since the entry was inserted.
        """
        # Stat in inode order as optimisation for at least linux.
        def inode_order(path_and_cache):
            return path_and_cache[1][1][3]
        for path, cache_val in sorted(self._cache.items(), key=inode_order):
            abspath = osutils.pathjoin(self.root, path)
            fp = self._fingerprint(abspath)
            self.stat_count += 1

            if not fp or cache_val[1] != fp:
                # not here or not a regular file anymore
                self.removed_count += 1
                self.needs_write = True
                del self._cache[path]

    def get_sha1(self, path, stat_value=None):
        """Return the sha1 of a file.
        """
        abspath = osutils.pathjoin(self.root, path)
        self.stat_count += 1
        file_fp = self._fingerprint(abspath, stat_value)

        if not file_fp:
            # not a regular file or not existing
            if path in self._cache:
                self.removed_count += 1
                self.needs_write = True
                del self._cache[path]
            return None

        if path in self._cache:
            cache_sha1, cache_fp = self._cache[path]
        else:
            cache_sha1, cache_fp = None, None

        if cache_fp == file_fp:
            self.hit_count += 1
            return cache_sha1

        self.miss_count += 1

        mode = file_fp[FP_MODE_COLUMN]
        if stat.S_ISREG(mode):
            if self._filter_provider is None:
                filters = []
            else:
                filters = self._filter_provider(path=path)
            digest = self._really_sha1_file(abspath, filters)
        elif stat.S_ISLNK(mode):
            target = osutils.readlink(abspath)
            digest = osutils.sha_string(target.encode('UTF-8'))
        else:
            raise errors.BzrError("file %r: unknown file stat mode: %o"
                                  % (abspath, mode))

        # window of 3 seconds to allow for 2s resolution on windows,
        # unsynchronized file servers, etc.
        cutoff = self._cutoff_time()
        if file_fp[FP_MTIME_COLUMN] >= cutoff \
                or file_fp[FP_CTIME_COLUMN] >= cutoff:
            # changed too recently; can't be cached.  we can
            # return the result and it could possibly be cached
            # next time.
            #
            # the point is that we only want to cache when we are sure that any
            # subsequent modifications of the file can be detected.  If a
            # modification neither changes the inode, the device, the size, nor
            # the mode, then we can only distinguish it by time; therefore we
            # need to let sufficient time elapse before we may cache this entry
            # again.  If we didn't do this, then, for example, a very quick 1
            # byte replacement in the file might go undetected.
            ## mutter('%r modified too recently; not caching', path)
            self.danger_count += 1
            if cache_fp:
                self.removed_count += 1
                self.needs_write = True
                del self._cache[path]
        else:
            # mutter('%r added to cache: now=%f, mtime=%d, ctime=%d',
            ##        path, time.time(), file_fp[FP_MTIME_COLUMN],
            # file_fp[FP_CTIME_COLUMN])
            self.update_count += 1
            self.needs_write = True
            self._cache[path] = (digest, file_fp)
        return digest

    def _really_sha1_file(self, abspath, filters):
        """Calculate the SHA1 of a file by reading the full text"""
        return _mod_filters.internal_size_sha_file_byname(abspath, filters)[1]

    def write(self):
        """Write contents of cache to file."""
        with atomicfile.AtomicFile(self.cache_file_name(), 'wb',
                                   new_mode=self._mode) as outf:
            outf.write(CACHE_HEADER)

            for path, c in self._cache.items():
                line_info = [path.encode('utf-8'), b'// ', c[0], b' ']
                line_info.append(b'%d %d %d %d %d %d' % c[1])
                line_info.append(b'\n')
                outf.write(b''.join(line_info))
            self.needs_write = False
            # mutter("write hash cache: %s hits=%d misses=%d stat=%d recent=%d updates=%d",
            #        self.cache_file_name(), self.hit_count, self.miss_count,
            # self.stat_count,
            # self.danger_count, self.update_count)

    def read(self):
        """Reinstate cache from file.

        Overwrites existing cache.

        If the cache file has the wrong version marker, this just clears
        the cache."""
        self._cache = {}

        fn = self.cache_file_name()
        try:
            inf = open(fn, 'rb', buffering=65000)
        except IOError as e:
            trace.mutter("failed to open %s: %s", fn, str(e))
            # better write it now so it is valid
            self.needs_write = True
            return

        with inf:
            hdr = inf.readline()
            if hdr != CACHE_HEADER:
                trace.mutter('cache header marker not found at top of %s;'
                             ' discarding cache', fn)
                self.needs_write = True
                return

            for l in inf:
                pos = l.index(b'// ')
                path = l[:pos].decode('utf-8')
                if path in self._cache:
                    trace.warning('duplicated path %r in cache' % path)
                    continue

                pos += 3
                fields = l[pos:].split(b' ')
                if len(fields) != 7:
                    trace.warning("bad line in hashcache: %r" % l)
                    continue

                sha1 = fields[0]
                if len(sha1) != 40:
                    trace.warning("bad sha1 in hashcache: %r" % sha1)
                    continue

                fp = tuple(map(int, fields[1:]))

                self._cache[path] = (sha1, fp)

        self.needs_write = False

    def _cutoff_time(self):
        """Return cutoff time.

        Files modified more recently than this time are at risk of being
        undetectably modified and so can't be cached.
        """
        return int(time.time()) - 3

    def _fingerprint(self, abspath, stat_value=None):
        if stat_value is None:
            try:
                stat_value = os.lstat(abspath)
            except OSError:
                # might be missing, etc
                return None
        if stat.S_ISDIR(stat_value.st_mode):
            return None
        # we discard any high precision because it's not reliable; perhaps we
        # could do better on some systems?
        return (stat_value.st_size, int(stat_value.st_mtime),
                int(stat_value.st_ctime), stat_value.st_ino,
                stat_value.st_dev, stat_value.st_mode)

File: bzrformats/bazaar/src/hashcache.rs
use crate::filters::{ContentFilter, ContentFilterProvider, ContentFilterStack};
use breezy_osutils::sha::sha_string;
use log::{debug, info};
use nix::sys::stat::SFlag;
use std::collections::HashMap;
use std::fs;
use std::fs::{File, Metadata, Permissions};
use std::io;
use std::io::prelude::*;
use std::io::BufReader;
use std::os::unix::fs::MetadataExt;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};
use tempfile::NamedTempFile;

/// TODO: Up-front, stat all files in order and remove those which are deleted or
/// out-of-date.  Don't actually re-read them until they're needed.  That ought
/// to bring all the inodes into core so that future stats to them are fast, and
/// it preserves the nice property that any caller will always get up-to-date
/// data except in unavoidable cases.

/// TODO: Perhaps return more details on the file to avoid statting it
/// again: nonexistent, file type, size, etc

const CACHE_HEADER: &[u8] = b"### bzr hashcache v5\n";

/// Cache for looking up file SHA-1.
///
/// Files are considered to match the cached value if the fingerprint
/// of the file has not changed.  This includes its mtime, ctime,
/// device number, inode number, and size.  This should catch
/// modifications or replacement of the file by a new one.
///
/// This may not catch modifications that do not change the file's
/// size and that occur within the resolution window of the
/// timestamps.  To handle this we specifically do not cache files
/// which have changed since the start of the present second, since
/// they could undetectably change again.
///
/// This scheme may fail if the machine's clock steps backwards.
/// Don't do that.
///
/// This does not canonicalize the paths passed in; that should be
/// done by the caller.
///
/// _cache
///     Indexed by path, points to a two-tuple of the SHA-1 of the file.
///     and its fingerprint.
///
/// stat_count
///     number of times files have been statted
///
/// hit_count
///     number of times files have been retrieved from the cache, avoiding a
///     re-read
///
/// miss_count
///     number of misses (times files have been completely re-read)

#[derive(Debug, PartialEq, Default, Clone)]
pub struct Fingerprint {
    pub size: u64,
    pub mtime: i64,
    pub ctime: i64,
    pub ino: u64,
    pub dev: u64,
    pub mode: u32,
}

impl From<Metadata> for Fingerprint {
    fn from(meta: Metadata) -> Fingerprint {
        Fingerprint {
            size: meta.size(),
            mtime: meta.mtime(),
            ctime: meta.ctime(),
            ino: meta.ino(),
            dev: meta.dev(),
            mode: meta.mode(),
        }
    }
}

const DEFAULT_CUTOFF_OFFSET: i64 = -3;

pub struct HashCache {
    root: PathBuf,
    hit_count: u32,
    miss_count: u32,
    stat_count: u32,
    danger_count: u32,
    removed_count: u32,
    update_count: u32,
    cache: HashMap<PathBuf, (String, Fingerprint)>,
    needs_write: bool,
    permissions: Option<Permissions>,
    cache_file_name: PathBuf,
    filter_provider: Option<Box<ContentFilterProvider>>,
    cutoff_offset: i64,
}

impl HashCache {
    /// Create a hash cache in base dir, and set the file mode to mode.
    ///
    /// Args:
    ///    content_filter_provider: a function that takes a
    ///       path (relative to the top of the tree) and a file-id as
    ///       parameters and returns a stack of ContentFilters.
    ///       If None, no content filtering is performed.
    pub fn new(
        root: &Path,
        cache_file_name: &Path,
        permissions: Option<Permissions>,
        content_filter_provider: Option<Box<ContentFilterProvider>>,
    ) -> Self {
        HashCache {
            root: root.to_path_buf(),
            hit_count: 0,
            miss_count: 0,
            stat_count: 0,
            danger_count: 0,
            removed_count: 0,
            update_count: 0,
            cache: HashMap::new(),
            needs_write: false,
            permissions,
            cache_file_name: cache_file_name.to_path_buf(),
            filter_provider: content_filter_provider,
            cutoff_offset: DEFAULT_CUTOFF_OFFSET,
        }
    }

    pub fn cache_file_name(&self) -> &Path {
        self.cache_file_name.as_path()
    }

    pub fn hit_count(&self) -> u32 {
        self.hit_count
    }

    pub fn miss_count(&self) -> u32 {
        self.miss_count
    }

    pub fn set_cutoff_offset(&mut self, offset: i64) {
        self.cutoff_offset = offset;
    }

    /// Discard all cached information.
    ///
    /// This does not reset the counters.
    pub fn clear(&mut self) {
        if !self.cache.is_empty() {
            self.needs_write = true;
            self.cache.clear();
        }
    }

    /// Scan all files and remove entries where the cache entry is obsolete.
    ///
    /// Obsolete entries are those where the file has been modified or deleted
    /// since the entry was inserted.
    pub fn scan(&mut self) {
        let mut keys_to_remove = Vec::new();
        let mut by_inode = self
            .cache
            .iter()
            .map(|(k, v)| (v.1.ino, k, v))
            .collect::<Vec<_>>();
        by_inode.sort_by_key(|x| x.0);
        for (_inode, path, cache_val) in by_inode {
            let abspath = self.root.join(path);
            let fp = self.fingerprint(abspath.as_ref(), None);
            self.stat_count += 1;

            if fp.is_none() || cache_val.1 != fp.unwrap() {
                // not here or not a regular file anymore
                self.removed_count += 1;
                self.needs_write = true;
                keys_to_remove.push(path.clone());
            }
        }
        for path in keys_to_remove {
            self.cache.remove(&path);
        }
    }

    pub fn get_sha1_by_fingerprint(
        &mut self,
        path: &Path,
        file_fp: &Fingerprint,
    ) -> io::Result<String> {
        let abspath = self.root.join(path);

        let (cache_sha1, cache_fp) = self
            .cache
            .get(path)
            .cloned()
            .unwrap_or((Default::default(), Default::default()));

        if cache_fp == *file_fp {
            self.hit_count += 1;
            Ok(cache_sha1)
        } else {
            self.miss_count += 1;

            match SFlag::from_bits_truncate(file_fp.mode as nix::libc::mode_t) {
                SFlag::S_IFREG => {
                    let filters: Box<dyn ContentFilter> =
                        if let Some(filter_provider) = self.filter_provider.as_ref() {
                            filter_provider(path, file_fp.ctime as u64)
                        } else {
                            Box::new(ContentFilterStack::new())
                        };
                    let digest = filters.sha1_file(&abspath)?;

                    // window of 3 seconds to allow for 2s resolution on windows,
                    // unsynchronized file servers, etc.
                    let cutoff = self.cutoff_time();
                    if file_fp.mtime >= cutoff || file_fp.ctime >= cutoff {
                        // changed too recently; can't be cached. we can
                        // return the result and it could possibly be cached
                        // next time.
                        //
                        // the point is that we only want to cache when we are sure that any
                        // subsequent modifications of the file can be detected. If a
                        // modification neither changes the inode, the device, the size, nor
                        // the mode, then we can only distinguish it by time; therefore we
                        // need to let sufficient time elapse before we may cache this entry
                        // again. If we didn't do this, then, for example, a very quick 1
                        // byte replacement in the file might go undetected.
                        self.danger_count += 1;
                        if self.cache.remove(path).is_some() {
                            self.removed_count += 1;
                            self.needs_write = true;
                        }
                    } else {
                        self.update_count += 1;
                        self.needs_write = true;
                        self.cache
                            .insert(path.to_owned(), (digest.clone(), file_fp.clone()));
                    }

                    Ok(digest)
                }
                SFlag::S_IFLNK => {
                    let target = fs::read_link(&abspath)?;
                    let digest = sha_string(target.to_string_lossy().as_bytes());
                    self.cache
                        .insert(path.to_owned(), (digest.clone(), file_fp.clone()));
                    self.update_count += 1;
                    self.needs_write = true;
                    Ok(digest)
                }
                _ => Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("unknown file stat mode: {:o}", file_fp.mode),
                )),
            }
        }
    }

    /// Return the SHA-1 of the file at path.
    pub fn get_sha1(
        &mut self,
        path: &Path,
        stat_value: Option<Metadata>,
    ) -> io::Result<Option<String>> {
        let abspath = self.root.join(path);
        self.stat_count += 1;
        let file_fp = self.fingerprint(abspath.as_ref(), stat_value);

        if file_fp.is_none() {
            // not a regular file or not existing
            if self.cache.remove(path).is_some() {
                self.removed_count += 1;
                self.needs_write = true;
            }

            Ok(None)
        } else {
            Ok(Some(self.get_sha1_by_fingerprint(path, &file_fp.unwrap())?))
        }
    }

    /// Write contents of cache to file.
    pub fn write(&mut self) -> Result<(), std::io::Error> {
        let mut outf = NamedTempFile::new_in(self.cache_file_name.parent().unwrap())?;
        if let Some(permissions) = self.permissions.clone() {
            outf.as_file().set_permissions(permissions)?;
        }
        outf.write_all(CACHE_HEADER)?;
        for (path, c) in &self.cache {
            let mut line_info: Vec<u8> = Vec::new();
            line_info.extend_from_slice(path.to_str().unwrap().as_bytes());
            line_info.extend_from_slice(b"// ");
            line_info.extend_from_slice(c.0.as_bytes());
            line_info.push(b' ');
            let fp = &c.1;
            write!(
                &mut line_info,
                "{} {} {} {} {} {}",
                fp.size, fp.mtime, fp.ctime, fp.ino, fp.dev, fp.mode
            )?;
            line_info.push(b'\n');
            outf.write_all(&line_info)?;
        }
        outf.persist(self.cache_file_name())?;
        self.needs_write = false;
        debug!(
            "write hash cache: {} hits={} misses={} stat={} recent={} updates={}",
            self.cache_file_name().display(),
            self.hit_count,
            self.miss_count,
            self.stat_count,
            self.danger_count,
            self.update_count
        );
        Ok(())
    }

    /// Reinstate cache from file.
    ///
    /// Overwrites existing cache.
    ///
    /// If the cache file has the wrong version marker, this just clears
    /// the cache.
    pub fn read(&mut self) -> Result<(), std::io::Error> {
        self.cache = HashMap::new();
        let file = File::open(self.cache_file_name());
        if file.is_err() {
            debug!(
                "failed to open {}: {}",
                self.cache_file_name().display(),
                file.err().unwrap()
            );
            self.needs_write = true;
            return Ok(());
        }
        let file = file.unwrap();
        let reader = BufReader::with_capacity(65000, file);
        let mut lines = reader.lines();
        if let Some(header) = lines.next() {
            if header?.as_bytes().eq(CACHE_HEADER) {
                self.needs_write = true;
                return Err(std::io::Error::new(
                    std::io::ErrorKind::InvalidData,
                    format!(
                        "cache header marker not found at top of {}; discarding cache",
                        self.cache_file_name().display()
                    ),
                ));
            }
        } else {
            self.needs_write = true;
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "error reading cache file header".to_string(),
            ));
        }
        for line in lines {
            let line = line?;
            let pos = line.find("// ").unwrap();
            let path = PathBuf::from(&line[..pos]);
            if self.cache.contains_key(&path) {
                info!("duplicated path {} in cache", path.display());
                continue;
            }
            let pos = pos + 3;
            let fields = line[pos..].split(' ').collect::<Vec<_>>();
            if fields.len() != 7 {
                info!("bad line in hashcache: {}", line);
                continue;
            }
            let sha1 = fields[0].to_owned();
            if sha1.len() != 40 {
                info!("bad sha1 in hashcache: {}", sha1);
                continue;
            }
            let fp = Fingerprint {
                size: fields[1].parse::<u64>().unwrap(),
                mtime: fields[2].parse::<i64>().unwrap(),
                ctime: fields[3].parse::<i64>().unwrap(),
                ino: fields[4].parse::<u64>().unwrap(),
                dev: fields[5].parse::<u64>().unwrap(),
                mode: fields[6].parse::<u32>().unwrap(),
            };
            self.cache.insert(path, (sha1, fp));
        }
        self.needs_write = false;
        Ok(())
    }

    pub fn needs_write(&self) -> bool {
        self.needs_write
    }

    /// Return cutoff time.
    ///
    /// Files modified more recently than this time are at risk of being
    /// undetectably modified and so can't be cached.
    pub fn cutoff_time(&self) -> i64 {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64
            + self.cutoff_offset
    }

    pub fn fingerprint(&self, abspath: &Path, stat_value: Option<Metadata>) -> Option<Fingerprint> {
        let stat_value = match stat_value {
            Some(s) => s,
            None => match fs::symlink_metadata(abspath) {
                Ok(s) => s,
                Err(_) => return None,
            },
        };
        if stat_value.is_dir() {
            return None;
        }
        Some(stat_value.into())
    }
}
